routing/fake_graph.hpp

```diff
@@ -0,0 +1,148 @@
+    {
+      m_realToFake[real].insert(newSegment);
+      m_fakeToReal[newSegment] = real;
+    }
+  }
+
+  // Merges |rhs| into this.
+  void Append(FakeGraph const & rhs)
+  {
+    auto IsIntersectionEmpty = [](std::map<SegmentType, VertexType> const & lhs,
+                                  std::map<SegmentType, VertexType> const & rhs) {
+      auto const counter =
+          std::set_intersection(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), CounterIterator(),
+                                my::LessBy(&std::pair<SegmentType, VertexType>::first));
+      return counter.GetCount() == 0;
+    };
+
+    CHECK(IsIntersectionEmpty(m_segmentToVertex, rhs.m_segmentToVertex),
+          ("Fake segments are not unique"));
+
+    m_segmentToVertex.insert(rhs.m_segmentToVertex.begin(), rhs.m_segmentToVertex.end());
+    m_vertexToSegment.insert(rhs.m_vertexToSegment.begin(), rhs.m_vertexToSegment.end());
+
+    for (auto const & kv : rhs.m_outgoing)
+      m_outgoing[kv.first].insert(kv.second.begin(), kv.second.end());
+
+    for (auto const & kv : rhs.m_ingoing)
+      m_ingoing[kv.first].insert(kv.second.begin(), kv.second.end());
+
+    for (auto const & kv : rhs.m_realToFake)
+      m_realToFake[kv.first].insert(kv.second.begin(), kv.second.end());
+
+    m_fakeToReal.insert(rhs.m_fakeToReal.begin(), rhs.m_fakeToReal.end());
+  }
+
+  // Returns Vertex which corresponds |segment|. Segment must be a part of the fake graph.
+  VertexType const & GetVertex(SegmentType const & segment) const
+  {
+    auto const it = m_segmentToVertex.find(segment);
+    CHECK(it != m_segmentToVertex.end(), ("Vertex for invalid fake segment requested."));
+    return it->second;
+  }
+
+  // Returns outgoing/ingoing edges set for specified segment.
+  std::set<SegmentType> const & GetEdges(SegmentType const & segment, bool isOutgoing) const
+  {
+    auto const & adjacentEdges = isOutgoing ? m_outgoing : m_ingoing;
+
+    auto const it = adjacentEdges.find(segment);
+    if (it != adjacentEdges.end())
+      return it->second;
+
+    return kEmptySet;
+  }
+
+  size_t GetSize() const { return m_segmentToVertex.size(); }
+
+  std::set<SegmentType> const & GetFake(RealType const & real) const
+  {
+    auto const it = m_realToFake.find(real);
+    if (it != m_realToFake.end())
+      return it->second;
+
+    return kEmptySet;
+  }
+
+  bool FindReal(SegmentType const & fake, RealType & real) const
+  {
+    auto const it = m_fakeToReal.find(fake);
+    if (it == m_fakeToReal.end())
+      return false;
+
+    real = it->second;
+    return true;
+  }
+
+  bool FindSegment(VertexType const & vertex, SegmentType & segment) const
+  {
+    auto const it = m_vertexToSegment.find(vertex);
+    if (it == m_vertexToSegment.end())
+      return false;
+
+    segment = it->second;
+    return true;
+  }
+
+  std::map<SegmentType, RealType> const & GetFakeToReal() const { return m_fakeToReal; }
+
+private:
+  // Key is fake segment, value is set of outgoing fake segments
```

Comment:
Let's add dots at the end of the sentences.

Reply:
done